File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *
import string

def timeToMin(s):
    hour,minute = map(int,s.split(":"))
    
    return  hour * 60 + minute


def solution(fees, records):
    
    
    defaultMin,defaultCost,unitMin,unitCost = fees
    
    _hash = defaultdict(int)  # 시간 : 입차 시간
    _index = defaultdict(int) # 시간 : index
    
    total_time = [] # 누적시간 
    answer = []
    
    for i,record in enumerate(records):
        time,car,state = record.split(" ")
        
        if state == "IN":
            if not(car in _index): # 처음 들어온 차면
                _index[car] = len(total_time) # 해당 차에대한 누저시간 인덱스를 맵핑
                total_time.append(0) # 해당 차에댄한 누적시간 생성 
            
            _hash[car] = timeToMin(time) # 입차 시간 저장
            
        else:
            total_time[_index[car]] += timeToMin(time) - _hash[car] # 현재시간 - 입차 시간
            del _hash[car] # 해당 차 지움
    
    
    for key,val in _hash.items(): # 남아있는 차는 23:39분 기준 출고 
        total_time[_index[key]] += timeToMin("23:59") - _hash[key]
    
   
    
    for key in sorted(_index.keys()): # 차량번호를 기준으로 오름 차순 
        totalMin =  total_time[_index[key]] # 누적시간 
        
        if totalMin <= defaultMin:
            answer.append(defaultCost)
        
        else:
            answer.append(defaultCost + int(ceil((totalMin - defaultMin)  / unitMin)) * unitCost) # 기본시간보다 크면 , 올림 처리 

            

        
         
        
    
    return answer

File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.swift
import Foundation

func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    var timeDict: Dictionary<String, Int> = [:] // 시간 기록 
    var parkingLot: Dictionary<String, String> = [:] //
    records.forEach {
        let map = $0.split(separator: " ").map{ String($0) }
        if parkingLot[map[1]] == nil { //현재 입차 안되어 있으면 
            parkingLot[map[1]] = map[0] // 입차 [차번호:시간]
        } else { // 없으면 출차 
            timeDict[map[1], default: 0] += parkTime(in: parkingLot[map[1]]!, out: map[0]) 
            // 주차 시간 기록 
            parkingLot[map[1]] = nil // 출차 
        }
    }
    parkingLot.forEach { // 아직 출차하지 않은 차 출차 

        timeDict[$0.key, default: 0] += parkTime(in: $0.value, out: "23:59")
    }
    var cars: [String] = []
    records.forEach { // 차 번호만 추출
        let carNumber = $0.split(separator: " ").map{ String($0) }[1]
        if !cars.contains(carNumber) { cars.append(carNumber) }
    }

    return cars.sorted(by: <).map {
        return fee(time: timeDict[$0]!, fees: fees)
    }
}

private func parkTime(in entrance: String, out exit: String) -> Int {
    let inTime = entrance.split(separator: ":").map{ String($0) }.reduce(0){ $0 * 60 + Int($1)! }
    let outTime = exit.split(separator: ":").map{ String($0) }.reduce(0){ $0 * 60 + Int($1)! }
    return outTime - inTime
}

private func fee(time: Int, fees: [Int]) -> Int {
    return time <= fees[0] ? fees[1] : fees[1] + Int(ceil(Double(time - fees[0])/Double(fees[2])) * Double(fees[3]))
}
